All Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/14210660725
BH post commit: https://github.com/tenstorrent/tt-metal/actions/runs/14210662583
Model regression: https://github.com/tenstorrent/tt-metal/actions/runs/14210664609 `AssertionError: Some model(s) Inference Time (sec) are too slow, see above for details on slow models: [('state-spaces/mamba-2.8b_mode-decode_layers-64_seqlen-1', 0.1192, 0.11)]` (possibly a spurious failure?)
Other perf regression tests appear to be failing on `main`...

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -196,10 +196,10 @@ add_compile_options(
     -Wreturn-type
     -Wswitch
     -Wuninitialized
-    -mavx2
     -fPIC
     -fvisibility-inlines-hidden
     -fno-lto # FIXME: This seems to be here for ttnn; it should go to TTNN, then.
+    "$<$<STREQUAL:${CMAKE_SYSTEM_PROCESSOR},x86_64>:-mavx2>"
     "$<$<CXX_COMPILER_ID:Clang>:-Wsometimes-uninitialized>"
     "$<$<CXX_COMPILER_ID:Clang>:-Wno-c++11-narrowing>"
     "$<$<CXX_COMPILER_ID:Clang>:-Wno-error=local-type-template-args>"

--- a/tt_metal/impl/data_format/bfloat4.cpp
+++ b/tt_metal/impl/data_format/bfloat4.cpp
@@ -2,8 +2,6 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include <emmintrin.h>
-#include <immintrin.h>
 #include <math.h>
 #include <tt-metalium/bfloat4.hpp>
 #include <tt_stl/span.hpp>
@@ -11,6 +9,7 @@
 #include <functional>
 #include <random>
 #include <vector>
+#include <simde/x86/avx2.h>
 
 #include "assert.hpp"
 #include "blockfloat_common.hpp"
@@ -73,14 +72,14 @@ std::vector<float> unpack_bfp4_tiles_into_float_vec(
     const std::vector<uint32_t> mask_vec1 = {0xf0000, 0xf00000, 0xf000000, 0xf0000000};
     const std::vector<uint32_t> shift_vec0 = {0, 4, 8, 12};
     const std::vector<uint32_t> shift_vec1 = {16, 20, 24, 28};
-    const __m128i mask0 = _mm_loadu_si128(reinterpret_cast<const __m128i*>(mask_vec0.data()));
-    const __m128i mask1 = _mm_loadu_si128(reinterpret_cast<const __m128i*>(mask_vec1.data()));
-    const __m128i shift0 = _mm_loadu_si128(reinterpret_cast<const __m128i*>(shift_vec0.data()));
-    const __m128i shift1 = _mm_loadu_si128(reinterpret_cast<const __m128i*>(shift_vec1.data()));
-    __m256i rebias_offset = _mm256_setzero_si256();
+    const simde__m128i mask0 = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(mask_vec0.data()));
+    const simde__m128i mask1 = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(mask_vec1.data()));
+    const simde__m128i shift0 = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(shift_vec0.data()));
+    const simde__m128i shift1 = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(shift_vec1.data()));
+    simde__m256i rebias_offset = simde_mm256_setzero_si256();
     if (is_exp_a) {
         rebias_offset =
-            _mm256_set1_epi32(-112);  // This rebias offset must be added if we are working with BFP8 format.
+            simde_mm256_set1_epi32(-112);  // This rebias offset must be added if we are working with BFP8 format.
     }
     uint32_t exp_word, sub_word_index;
 
@@ -100,8 +99,8 @@ std::vector<float> unpack_bfp4_tiles_into_float_vec(
                 for (int i = 0; i < subtile_rows; ++i) {
                     subtile_r = tr * subtile_rows + i;
                     for (int j = 0; j < subtile_cols; j += 2 * num_elements_in_dword) {
-                        __m256i mask_denormal0 = _mm256_setzero_si256();
-                        __m256i mask_denormal1 = _mm256_setzero_si256();
+                        simde__m256i mask_denormal0 = simde_mm256_setzero_si256();
+                        simde__m256i mask_denormal1 = simde_mm256_setzero_si256();
                         subtile_c = tc * subtile_cols + j;
                         data_index =
                             (tr * (subtiles_in_tile_col * face_HW / num_elements_in_dword) +
@@ -120,108 +119,116 @@ std::vector<float> unpack_bfp4_tiles_into_float_vec(
                         sub_word_index = ((tile_and_data_index - num_exponent_words_skip) >> data_dwords_per_exp_log2) &
                                          exp_bit_mask;  // Extract the byte in which the shared exponent is stored. Each
                                                         // byte is shared amongst 16 datums.
-                        __m256i exp_vector0 =
-                            _mm256_set1_epi32(get_byte(exp_word, sub_word_index));  // Replicate exp scalar in a vector
-                        __m256i exp_vector1 = exp_vector0;
+                        simde__m256i exp_vector0 = simde_mm256_set1_epi32(
+                            get_byte(exp_word, sub_word_index));  // Replicate exp scalar in a vector
+                        simde__m256i exp_vector1 = exp_vector0;
 
                         // Take 2 uint32_t values. These are 16 BFP4 values
                         // Replicate each uint32 value 8 times - one for each bfp4 value
                         uint32_t first_val = bfp_tiles.at(num_exp_words + tile_and_data_index);
-                        __m128i first0 = _mm_set1_epi32(first_val);
-                        __m128i first1 = _mm_set1_epi32(first_val);
+                        simde__m128i first0 = simde_mm_set1_epi32(first_val);
+                        simde__m128i first1 = simde_mm_set1_epi32(first_val);
                         uint32_t second_val = bfp_tiles.at(num_exp_words + tile_and_data_index + 1);
-                        __m128i second0 = _mm_set1_epi32(second_val);
-                        __m128i second1 = _mm_set1_epi32(second_val);
+                        simde__m128i second0 = simde_mm_set1_epi32(second_val);
+                        simde__m128i second1 = simde_mm_set1_epi32(second_val);
 
-                        first0 = _mm_srlv_epi32(
-                            _mm_and_si128(first0, mask0), shift0);  // Extract each BFP4 from the first0 uint32_t
-                        first1 = _mm_srlv_epi32(
-                            _mm_and_si128(first1, mask1), shift1);  // Extract each BFP4 from the first1 uint32_t
-                        second0 = _mm_srlv_epi32(
-                            _mm_and_si128(second0, mask0), shift0);  // Extract each BFP4 from the first0 uint32_t
-                        second1 = _mm_srlv_epi32(
-                            _mm_and_si128(second1, mask1), shift1);  // Extract each BFP4 from the first1 uint32_t
-                        __m256i combined0 = _mm256_set_m128i(first1, first0);    // Concatenate 2 128 vectors to 1 256
-                        __m256i combined1 = _mm256_set_m128i(second1, second0);  // Concatenate 2 128 vectors to 1 256
+                        first0 = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(first0, mask0), shift0);  // Extract each BFP4 from the first0 uint32_t
+                        first1 = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(first1, mask1), shift1);  // Extract each BFP4 from the first1 uint32_t
+                        second0 = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(second0, mask0), shift0);  // Extract each BFP4 from the first0 uint32_t
+                        second1 = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(second1, mask1), shift1);  // Extract each BFP4 from the first1 uint32_t
+                        simde__m256i combined0 =
+                            simde_mm256_set_m128i(first1, first0);  // Concatenate 2 128 vectors to 1 256
+                        simde__m256i combined1 =
+                            simde_mm256_set_m128i(second1, second0);  // Concatenate 2 128 vectors to 1 256
 
                         // Extract sign and mantissa (expo extracted above)
-                        __m256i sign0 = _mm256_srl_epi32(combined0, _mm_set_epi64x(0, 3));
-                        __m256i man0 = _mm256_and_si256(combined0, _mm256_set1_epi32(0x7));
-                        __m256i sign1 = _mm256_srl_epi32(combined1, _mm_set_epi64x(0, 3));
-                        __m256i man1 = _mm256_and_si256(combined1, _mm256_set1_epi32(0x7));
+                        simde__m256i sign0 = simde_mm256_srl_epi32(combined0, simde_mm_set_epi64x(0, 3));
+                        simde__m256i man0 = simde_mm256_and_si256(combined0, simde_mm256_set1_epi32(0x7));
+                        simde__m256i sign1 = simde_mm256_srl_epi32(combined1, simde_mm_set_epi64x(0, 3));
+                        simde__m256i man1 = simde_mm256_and_si256(combined1, simde_mm256_set1_epi32(0x7));
 
                         for (int i = 0; i < 2; i++) {
-                            __m256i shift_cnt = _mm256_setzero_si256();  // Initialize shift amount per datum to 0. This
-                                                                         // is incremented below.
-                            __m256i man_shifted = i == 0 ? man0 : man1;  // Initialize updated mantissa
-                            __m256i select_mask = _mm256_cmpeq_epi32(
+                            simde__m256i shift_cnt = simde_mm256_setzero_si256();  // Initialize shift amount per datum
+                                                                                   // to 0. This is incremented below.
+                            simde__m256i man_shifted = i == 0 ? man0 : man1;       // Initialize updated mantissa
+                            simde__m256i select_mask = simde_mm256_cmpeq_epi32(
                                 man_shifted,
                                 shift_cnt);  // This mask is used to set mantissa values to 0, if they start at 0.
                             for (int shift_val = 0; shift_val < 3; shift_val++) {
                                 // Shift each mantissa and update the corresponding shift_cnt until the 3rd bit of the 8
                                 // bit data is set.
-                                __m256i shift_mask = _mm256_or_si256(
-                                    _mm256_cmpgt_epi32(man_shifted, _mm256_set1_epi32(0x4)),
-                                    _mm256_cmpeq_epi32(
+                                simde__m256i shift_mask = simde_mm256_or_si256(
+                                    simde_mm256_cmpgt_epi32(man_shifted, simde_mm256_set1_epi32(0x4)),
+                                    simde_mm256_cmpeq_epi32(
                                         man_shifted,
-                                        _mm256_set1_epi32(0x4)));  // If the 6th bit is set, propagate the current
-                                                                   // mantissa value. Else take the left shifted value
-                                man_shifted = _mm256_blendv_epi8(
-                                    _mm256_sll_epi32(man_shifted, _mm_set_epi64x(0, 1)), man_shifted, shift_mask);
-                                shift_cnt = _mm256_blendv_epi8(_mm256_set1_epi32(shift_val + 1), shift_cnt, shift_mask);
+                                        simde_mm256_set1_epi32(
+                                            0x4)));  // If the 6th bit is set, propagate the current
+                                                     // mantissa value. Else take the left shifted value
+                                man_shifted = simde_mm256_blendv_epi8(
+                                    simde_mm256_sll_epi32(man_shifted, simde_mm_set_epi64x(0, 1)),
+                                    man_shifted,
+                                    shift_mask);
+                                shift_cnt = simde_mm256_blendv_epi8(
+                                    simde_mm256_set1_epi32(shift_val + 1), shift_cnt, shift_mask);
                             }
-                            man_shifted = _mm256_and_si256(
-                                _mm256_sll_epi32(man_shifted, _mm_set_epi64x(0, 1)),
-                                _mm256_set1_epi32(
+                            man_shifted = simde_mm256_and_si256(
+                                simde_mm256_sll_epi32(man_shifted, simde_mm_set_epi64x(0, 1)),
+                                simde_mm256_set1_epi32(
                                     0x7));  // One more shift to clear 3rd bit; Mask with 3bits for mantissa
                             if (i == 0) {
-                                man0 = _mm256_blendv_epi8(
+                                man0 = simde_mm256_blendv_epi8(
                                     man_shifted, man0, select_mask);  // Choose new mantissa or keep old mantissa based
                                                                       // on 0 initial condition.
                                 // Flush denormals to zero
                                 // Check if shift > exp and mantissa is not zero
-                                __m256i mask_shift_gt_exp = _mm256_cmpgt_epi32(shift_cnt, exp_vector0);
-                                __m256i mask_nonzero_mantissa = _mm256_xor_si256(select_mask, _mm256_set1_epi32(-1));
-                                mask_denormal0 = _mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
+                                simde__m256i mask_shift_gt_exp = simde_mm256_cmpgt_epi32(shift_cnt, exp_vector0);
+                                simde__m256i mask_nonzero_mantissa = simde_mm256_xor_si256(select_mask, simde_mm256_set1_epi32(-1));
+                                mask_denormal0 = simde_mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
 
-                                exp_vector0 = _mm256_blendv_epi8(
-                                    _mm256_sub_epi32(exp_vector0, _mm256_add_epi32(rebias_offset, shift_cnt)),
-                                    _mm256_setzero_si256(),
+                                exp_vector0 = simde_mm256_blendv_epi8(
+                                    simde_mm256_sub_epi32(exp_vector0, simde_mm256_add_epi32(rebias_offset, shift_cnt)),
+                                    simde_mm256_setzero_si256(),
                                     select_mask);  // Choose new (rebiased exponent) or keep previous exponent based on
                                                    // mantissa intiial condition
                             } else {
-                                man1 = _mm256_blendv_epi8(
+                                man1 = simde_mm256_blendv_epi8(
                                     man_shifted, man1, select_mask);  // Choose new mantissa or keep old mantissa based
                                                                       // on 0 initial condition.
                                 // Flush denormals to zero
                                 // Check if shift > exp and mantissa is not zero
-                                __m256i mask_shift_gt_exp = _mm256_cmpgt_epi32(shift_cnt, exp_vector1);
-                                __m256i mask_nonzero_mantissa = _mm256_xor_si256(select_mask, _mm256_set1_epi32(-1));
-                                mask_denormal1 = _mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
-                                exp_vector1 = _mm256_blendv_epi8(
-                                    _mm256_sub_epi32(exp_vector1, _mm256_add_epi32(rebias_offset, shift_cnt)),
-                                    _mm256_setzero_si256(),
+                                simde__m256i mask_shift_gt_exp = simde_mm256_cmpgt_epi32(shift_cnt, exp_vector1);
+                                simde__m256i mask_nonzero_mantissa = simde_mm256_xor_si256(select_mask, simde_mm256_set1_epi32(-1));
+                                mask_denormal1 = simde_mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
+                                exp_vector1 = simde_mm256_blendv_epi8(
+                                    simde_mm256_sub_epi32(exp_vector1, simde_mm256_add_epi32(rebias_offset, shift_cnt)),
+                                    simde_mm256_setzero_si256(),
                                     select_mask);  // Choose new (rebiased exponent) or keep previous exponent based on
                                                    // mantissa intiial condition
                             }
                         }
 
-                        sign0 = _mm256_sll_epi32(sign0, _mm_set_epi64x(0, 31));              // Shift sign
-                        sign1 = _mm256_sll_epi32(sign1, _mm_set_epi64x(0, 31));              // Shift sign
-                        exp_vector0 = _mm256_sll_epi32(exp_vector0, _mm_set_epi64x(0, 23));  // Shift exp
-                        exp_vector1 = _mm256_sll_epi32(exp_vector1, _mm_set_epi64x(0, 23));  // Shift exp
-                        man0 = _mm256_sll_epi32(man0, _mm_set_epi64x(0, 20));                // Shift mantissa
-                        man0 = _mm256_or_si256(
+                        sign0 = simde_mm256_sll_epi32(sign0, simde_mm_set_epi64x(0, 31));              // Shift sign
+                        sign1 = simde_mm256_sll_epi32(sign1, simde_mm_set_epi64x(0, 31));              // Shift sign
+                        exp_vector0 = simde_mm256_sll_epi32(exp_vector0, simde_mm_set_epi64x(0, 23));  // Shift exp
+                        exp_vector1 = simde_mm256_sll_epi32(exp_vector1, simde_mm_set_epi64x(0, 23));  // Shift exp
+                        man0 = simde_mm256_sll_epi32(man0, simde_mm_set_epi64x(0, 20));                // Shift mantissa
+                        man0 = simde_mm256_or_si256(
                             sign0,
-                            _mm256_or_si256(exp_vector0, man0));  // Store final value in mantissa register and save
-                        man1 = _mm256_sll_epi32(man1, _mm_set_epi64x(0, 20));  // Shift mantissa
-                        man1 = _mm256_or_si256(
+                            simde_mm256_or_si256(
+                                exp_vector0, man0));  // Store final value in mantissa register and save
+                        man1 = simde_mm256_sll_epi32(man1, simde_mm_set_epi64x(0, 20));  // Shift mantissa
+                        man1 = simde_mm256_or_si256(
                             sign1,
-                            _mm256_or_si256(exp_vector1, man1));  // Store final value in mantissa register and save
+                            simde_mm256_or_si256(
+                                exp_vector1, man1));  // Store final value in mantissa register and save
 
                         // Zero out lanes where mask_denormal is true
-                        man0 = _mm256_blendv_epi8(man0, _mm256_setzero_si256(), mask_denormal0);
-                        man1 = _mm256_blendv_epi8(man1, _mm256_setzero_si256(), mask_denormal1);
+                        man0 = simde_mm256_blendv_epi8(man0, simde_mm256_setzero_si256(), mask_denormal0);
+                        man1 = simde_mm256_blendv_epi8(man1, simde_mm256_setzero_si256(), mask_denormal1);
                         uint32_t float_data_index;
                         if (row_major_output) {
                             float_data_index = subtile_c + (tile_W * subtile_r) + (tile_index * num_float_in_tile);
@@ -229,9 +236,9 @@ std::vector<float> unpack_bfp4_tiles_into_float_vec(
                             float_data_index = fp32_element_index;
                             fp32_element_index += 2 * num_elements_in_dword;
                         }
-                        _mm256_storeu_ps(&float_vec[float_data_index], _mm256_castsi256_ps(man0));
-                        _mm256_storeu_ps(
-                            &float_vec[float_data_index + num_elements_in_dword], _mm256_castsi256_ps(man1));
+                        simde_mm256_storeu_ps(&float_vec[float_data_index], simde_mm256_castsi256_ps(man0));
+                        simde_mm256_storeu_ps(
+                            &float_vec[float_data_index + num_elements_in_dword], simde_mm256_castsi256_ps(man1));
                     }
                 }
             }

--- a/tt_metal/impl/data_format/bfloat8.cpp
+++ b/tt_metal/impl/data_format/bfloat8.cpp
@@ -2,14 +2,13 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include <emmintrin.h>
-#include <immintrin.h>
 #include <tt-metalium/bfloat8.hpp>
 #include <tt_stl/span.hpp>
 #include <array>
 #include <functional>
 #include <random>
 #include <vector>
+#include <simde/x86/avx2.h>
 
 #include "assert.hpp"
 #include "blockfloat_common.hpp"
@@ -68,12 +67,12 @@ std::vector<float> unpack_bfp8_tiles_into_float_vec(
     int subtile_c;
     const std::vector<uint32_t> mask_vec = {0xff, 0xff00, 0xff0000, 0xff000000};
     const std::vector<uint32_t> shift_vec = {0, 8, 16, 24};
-    const __m128i mask = _mm_loadu_si128(reinterpret_cast<const __m128i*>(mask_vec.data()));
-    const __m128i shift = _mm_loadu_si128(reinterpret_cast<const __m128i*>(shift_vec.data()));
-    __m256i rebias_offset = _mm256_setzero_si256();
+    const simde__m128i mask = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(mask_vec.data()));
+    const simde__m128i shift = simde_mm_loadu_si128(reinterpret_cast<const simde__m128i*>(shift_vec.data()));
+    simde__m256i rebias_offset = simde_mm256_setzero_si256();
     if (is_exp_a) {
         rebias_offset =
-            _mm256_set1_epi32(-112);  // This rebias offset must be added if we are working with BFP8 format.
+            simde_mm256_set1_epi32(-112);  // This rebias offset must be added if we are working with BFP8 format.
     }
     uint32_t exp_word, sub_word_index;
 
@@ -102,67 +101,70 @@ std::vector<float> unpack_bfp8_tiles_into_float_vec(
                         sub_word_index = ((tile_and_data_index - num_exponent_words_skip) >> 2) &
                                          exp_bit_mask;  // Extract the byte in which the shared exponent is stored. Each
                                                         // byte is shared amongst 16 datums.
-                        __m256i exp_vector =
-                            _mm256_set1_epi32(get_byte(exp_word, sub_word_index));  // Replicate exp scalar in a vector
+                        simde__m256i exp_vector = simde_mm256_set1_epi32(
+                            get_byte(exp_word, sub_word_index));  // Replicate exp scalar in a vector
                         // Take 2 uint32_t values. These are 8 BFP8 values
-                        __m128i first = _mm_set1_epi32(bfp8_tiles.at(
+                        simde__m128i first = simde_mm_set1_epi32(bfp8_tiles.at(
                             num_exp_words +
                             tile_and_data_index));  // Replicate first uint32_t 4 times (one for each BFP8 value)
-                        __m128i second = _mm_set1_epi32(bfp8_tiles.at(
+                        simde__m128i second = simde_mm_set1_epi32(bfp8_tiles.at(
                             num_exp_words + tile_and_data_index + 1));  //  Replicate second uint32_t 4 times
-                        first = _mm_srlv_epi32(
-                            _mm_and_si128(first, mask), shift);  // Extract each BFP8 from the first uint32_t
-                        second = _mm_srlv_epi32(
-                            _mm_and_si128(second, mask), shift);  // Extract each BFP8 from the second uint32_t
-                        __m256i combined = _mm256_set_m128i(second, first);  // Concatenate 2 128 vectors to 1 256
+                        first = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(first, mask), shift);  // Extract each BFP8 from the first uint32_t
+                        second = simde_mm_srlv_epi32(
+                            simde_mm_and_si128(second, mask), shift);  // Extract each BFP8 from the second uint32_t
+                        simde__m256i combined =
+                            simde_mm256_set_m128i(second, first);  // Concatenate 2 128 vectors to 1 256
                         // Extract sign and mantissa (expo extracted above)
-                        __m256i sign = _mm256_srl_epi32(combined, _mm_set_epi64x(0, 7));
-                        __m256i man = _mm256_and_si256(combined, _mm256_set1_epi32(0x7f));
+                        simde__m256i sign = simde_mm256_srl_epi32(combined, simde_mm_set_epi64x(0, 7));
+                        simde__m256i man = simde_mm256_and_si256(combined, simde_mm256_set1_epi32(0x7f));
                         // Initialize shift amount per datum to 0. This is incremented below.
-                        __m256i shift_cnt = _mm256_setzero_si256();
-                        __m256i select_mask = _mm256_cmpeq_epi32(
+                        simde__m256i shift_cnt = simde_mm256_setzero_si256();
+                        simde__m256i select_mask = simde_mm256_cmpeq_epi32(
                             man, shift_cnt);  // This mask is used to set mantissa values to 0, if they start at 0.
-                        __m256i man_shifted = man;  // Initialize updated mantissa
+                        simde__m256i man_shifted = man;  // Initialize updated mantissa
                         for (int shift_val = 0; shift_val < 7; shift_val++) {
                             // Shift each mantissa and update the corresponding shift_cnt until the 6th bit of the 8 bit
                             // data is set.
-                            __m256i shift_mask = _mm256_or_si256(
-                                _mm256_cmpgt_epi32(man_shifted, _mm256_set1_epi32(0x40)),
-                                _mm256_cmpeq_epi32(
+                            simde__m256i shift_mask = simde_mm256_or_si256(
+                                simde_mm256_cmpgt_epi32(man_shifted, simde_mm256_set1_epi32(0x40)),
+                                simde_mm256_cmpeq_epi32(
                                     man_shifted,
-                                    _mm256_set1_epi32(0x40)));  // If the 6th bit is set, propagate the current mantissa
-                                                                // value. Else take the left shifted value
-                            man_shifted = _mm256_blendv_epi8(
-                                _mm256_sll_epi32(man_shifted, _mm_set_epi64x(0, 1)), man_shifted, shift_mask);
-                            shift_cnt = _mm256_blendv_epi8(_mm256_set1_epi32(shift_val + 1), shift_cnt, shift_mask);
+                                    simde_mm256_set1_epi32(0x40)));  // If the 6th bit is set, propagate the current
+                                                                     // mantissa value. Else take the left shifted value
+                            man_shifted = simde_mm256_blendv_epi8(
+                                simde_mm256_sll_epi32(man_shifted, simde_mm_set_epi64x(0, 1)), man_shifted, shift_mask);
+                            shift_cnt =
+                                simde_mm256_blendv_epi8(simde_mm256_set1_epi32(shift_val + 1), shift_cnt, shift_mask);
                         }
-                        man_shifted = _mm256_and_si256(
-                            _mm256_sll_epi32(man_shifted, _mm_set_epi64x(0, 1)),
-                            _mm256_set1_epi32(0x7f));  // One more shift to clear 6th bit
-                        man = _mm256_blendv_epi8(
+                        man_shifted = simde_mm256_and_si256(
+                            simde_mm256_sll_epi32(man_shifted, simde_mm_set_epi64x(0, 1)),
+                            simde_mm256_set1_epi32(0x7f));  // One more shift to clear 6th bit
+                        man = simde_mm256_blendv_epi8(
                             man_shifted,
                             man,
                             select_mask);  // Choose new mantissa or keep old mantissa based on 0 initial condition.
                         // Flush denormals to zero
                         // Check if shift > exp and mantissa is not zero
-                        __m256i mask_shift_gt_exp = _mm256_cmpgt_epi32(shift_cnt, exp_vector);
-                        __m256i mask_nonzero_mantissa = _mm256_xor_si256(select_mask, _mm256_set1_epi32(-1));
-                        __m256i mask_denormal = _mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
+                        simde__m256i mask_shift_gt_exp = simde_mm256_cmpgt_epi32(shift_cnt, exp_vector);
+                        simde__m256i mask_nonzero_mantissa = simde_mm256_xor_si256(select_mask, simde_mm256_set1_epi32(-1));
+                        simde__m256i mask_denormal = simde_mm256_and_si256(mask_shift_gt_exp, mask_nonzero_mantissa);
 
-                        exp_vector = _mm256_blendv_epi8(
-                            _mm256_sub_epi32(exp_vector, _mm256_add_epi32(rebias_offset, shift_cnt)),
-                            _mm256_setzero_si256(),
+                        exp_vector = simde_mm256_blendv_epi8(
+                            simde_mm256_sub_epi32(exp_vector, simde_mm256_add_epi32(rebias_offset, shift_cnt)),
+                            simde_mm256_setzero_si256(),
                             select_mask);  // Choose new (rebiased exponent) or keep previous exponent based on mantissa
                                            // intiial condition
 
-                        sign = _mm256_sll_epi32(sign, _mm_set_epi64x(0, 31));              // Shift sign
-                        exp_vector = _mm256_sll_epi32(exp_vector, _mm_set_epi64x(0, 23));  // Shift exp
-                        man = _mm256_sll_epi32(man, _mm_set_epi64x(0, 16));                // Shift mantissa
-                        man = _mm256_or_si256(
-                            sign, _mm256_or_si256(exp_vector, man));  // Store final value in mantissa register and save
+                        sign = simde_mm256_sll_epi32(sign, simde_mm_set_epi64x(0, 31));              // Shift sign
+                        exp_vector = simde_mm256_sll_epi32(exp_vector, simde_mm_set_epi64x(0, 23));  // Shift exp
+                        man = simde_mm256_sll_epi32(man, simde_mm_set_epi64x(0, 16));                // Shift mantissa
+                        man = simde_mm256_or_si256(
+                            sign,
+                            simde_mm256_or_si256(exp_vector, man));  // Store final value in mantissa register and save
 
                         // Zero out lanes where mask_denormal is true
-                        man = _mm256_blendv_epi8(man, _mm256_setzero_si256(), mask_denormal);
+                        man = simde_mm256_blendv_epi8(man, simde_mm256_setzero_si256(), mask_denormal);
 
                         uint32_t float_data_index;
                         if (row_major_output) {
@@ -171,7 +173,7 @@ std::vector<float> unpack_bfp8_tiles_into_float_vec(
                             float_data_index = fp32_element_index;
                             fp32_element_index += 8;
                         }
-                        _mm256_storeu_ps(&float_vec[float_data_index], _mm256_castsi256_ps(man));
+                        simde_mm256_storeu_ps(&float_vec[float_data_index], simde_mm256_castsi256_ps(man));
                     }
                 }
             }

--- a/tt_metal/impl/dispatch/memcpy.hpp
+++ b/tt_metal/impl/dispatch/memcpy.hpp
@@ -5,12 +5,15 @@
 #pragma once
 
 #include <cstdint>
-#include <emmintrin.h>
 #include "assert.hpp"
 #include <tt_stl/aligned_allocator.hpp>
 #include <umd/device/device_api_metal.h>
 
 #include <tt-metalium/vector_aligned.hpp>
+
+#if defined(__x86_64__) || defined(__i386__)
+#include <emmintrin.h>
+#endif
 
 namespace tt::tt_metal {
 
@@ -18,6 +21,7 @@ namespace tt::tt_metal {
 // Benchmarked to be approximately 1.4x - 1.8x faster than std::memcpy
 // TODO: Revisit this w/ regard to possibly eliminating min sizes and orphan writes at the end
 // TODO: ditto alignment isues
+#if defined(__x86_64__) || defined(__i386__)
 template <bool debug_sync = false>
 static inline void memcpy_to_device(void* __restrict dst, const void* __restrict src, size_t n) {
     TT_ASSERT((uintptr_t)dst % MEMCPY_ALIGNMENT == 0);
@@ -96,5 +100,15 @@ static inline void memcpy_to_device(void* __restrict dst, const void* __restrict
         tt_driver_atomics::sfence();
     }
 }
+#else
+template <bool debug_sync = false>
+__attribute((nonnull(1, 2))) static inline void memcpy_to_device(
+    void* __restrict dst, const void* __restrict src, size_t n) {
+    memcpy(dst, src, n);
+    if constexpr (debug_sync) {
+        tt_driver_atomics::sfence();
+    }
+}
+#endif
 
 }  // namespace tt::tt_metal

